<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Promise-执行规则的完善</title>
</head>

<body>
  <script>
    // 判断变量否为function
    const isFunction = variable => typeof variable === 'function'
    // 定义Promise的三种状态常量
    const PENDING = 'PENDING'
    const FULFILLED = 'FULFILLED'
    const REJECTED = 'REJECTED'
    class MyPromise {
      constructor(handle) {
        if (!isFunction(handle)) {
          throw new Error('MyPromise must accept a function as a parameter')
        }
        // 添加状态
        this._status = PENDING
        // 添加状态
        this._value = undefined
        // 添加成功回调函数队列
        this._fulfilledQueues = []
        // 添加失败回调函数队列
        this._rejectedQueues = []
        // 执行handle (resolve,reject)=>{},参数：_resolve,_reject
        try {
          handle(this._resolve.bind(this), this._reject.bind(this))
        } catch (err) {
          this._reject(err)
        }
      }
      // 添加resovle时执行的函数 ----------改变status的状态和value的值
      _resolve(val) {
        const run = () => {
          if (this._status !== PENDING) return
          // 依次执行成功队列中的函数，并清空队列
          const runFulfilled = (value) => {
            let cb;
            while (cb = this._fulfilledQueues.shift()) {
              cb(value)
            }
          }
          // 依次执行失败队列中的函数，并清空队列
          const runRejected = (error) => {
            let cb;
            while (cb = this._rejectedQueues.shift()) {
              cb(error)
            }
          }
          /* 如果resolve的参数为Promise对象，则必须等待该Promise对象状态改变后,
            当前Promsie的状态才会改变，且状态取决于参数Promsie对象的状态
          */
          if (val instanceof MyPromise) {
            val.then(value => {
              this._value = value
              this._status = FULFILLED
              runFulfilled(value)
            }, err => {
              this._value = err
              this._status = REJECTED
              runRejected(err)
            })
          } else {
            this._value = val
            this._status = FULFILLED
            runFulfilled(val)
          }
        }
        // 为了支持同步的Promise，这里采用异步调用
        setTimeout(run, 0)
      }
      // 添加reject时执行的函数
      _reject(err) {
        if (this._status !== PENDING) return
        const run = () => {
          this._status = REJECTED
          this._value = err
          let cb;
          while (cb = this._rejectedQueues.shift()) {
            cb(err)
          }
        }
        setTimeout(run, 0)
      }
      // 添加then方法
      then(onFulfilled, onRejected) {
        const { _value, _status } = this
        // 返回一个新的Promise对象
        return new MyPromise((onFulfilledNext, onRejectedNext) => {
          // 封装一个成功时执行的函数
          let fulfilled = value => {
            try {
              if (!isFunction(onFulfilled)) {
                onFulfilledNext(value)
              } else {
                let res = onFulfilled(value);
                if (res instanceof MyPromise) {
                  // 如果当前回调函数返回MyPromise对象，必须等待其状态改变后在执行下一个回调
                  res.then(onFulfilledNext, onRejectedNext)
                } else {
                  //否则会将返回结果直接作为参数，传入下一个then的回调函数，并立即执行下一个then的回调函数
                  onFulfilledNext(res)
                }
              }
            } catch (err) {
              // 如果函数执行出错，新的Promise对象的状态为失败
              onRejectedNext(err)
            }
          }
          // 封装一个失败时执行的函数
          let rejected = error => {
            try {
              if (!isFunction(onRejected)) {
                onRejectedNext(error)
              } else {
                let res = onRejected(error);
                if (res instanceof MyPromise) {
                  // 如果当前回调函数返回MyPromise对象，必须等待其状态改变后在执行下一个回调
                  res.then(onFulfilledNext, onRejectedNext)
                } else {
                  //否则会将返回结果直接作为参数，传入下一个then的回调函数，并立即执行下一个then的回调函数
                  onFulfilledNext(res)
                }
              }
            } catch (err) {
              // 如果函数执行出错，新的Promise对象的状态为失败
              onRejectedNext(err)
            }
          }
          switch (_status) {
            // 当状态为pending时，将then方法回调函数加入执行队列等待执行
            case PENDING:
              this._fulfilledQueues.push(fulfilled)
              this._rejectedQueues.push(rejected)
              break
            case FULFILLED:
              fulfilled(_value)
              break
            case REJECTED:
              rejected(_value)
              break
          }
        })
      }
      // 添加catch方法
      catch(onRejected) {
        return this.then(undefined, onRejected)
      }
    }
  </script>
  <script>

    /*
      promise的执行规则：
      1.onFulfilled或onRejected 为函数,有两种情况；
        a.若x返回值若不为Promise对象，即x值是新的onFulfilled 或者 onRejected 函数的参数.
        b.若x返回值若为Promise对象，就会等待该 Promise 对象(即 x )的状态发生变化，才会被调用，并且新的 Promise 状态和 x 的状态相同.
     2.如果 onFulfilled 或者onRejected 抛出一个异常 e ，则 p2 必须变为失败（Rejected），并返回失败的值 e.
     3.如果 onFulfilled 不是函数且 p1 状态为成功（Fulfilled）， p2 必须变为成功（Fulfilled并返回p1成功的值.
     4.如果 onRejected  不是函数且 p1 状态为失败（Rejected），p2必须变为失败（Rejected） 并返回 p1 失败的值.
    */
    let p1, p2
    debugger
    // 例子1-a:
    p1 = new MyPromise((resolve, reject) => {
      resolve('成功1')
        console.log(111111111)
    })
    p2 = p1.then(res => {
      return new MyPromise((resolve, reject) => {
        resolve('这里返回一个Promise')
      })
    })
    p2.then(res => { console.log(res) })  //这里返回一个Promise

    // 例子1-b:
    // p1 = new MyPromise((resolve, reject) => {
    //   resolve('成功1')
    // })
    // p2 = p1.then(res => {
    //    return '这里返回一个普通值'
    // })
    // p2.then(res => { console.log(res) })  //这里返回一个普通值


    // // 例子2:
    // p1 = new MyPromise((resolve, reject) => {
    //   resolve('成功1')
    // })
    // p2 = p1.then(res => {
    //   throw new Error('这里抛出一个异常e')
    // })
    // p2.then(
    //   res => { console.log(res) },
    //   err => { console.log(err)
    // })  //这里抛出一个异常e

    // // 例子3:
    // p1 = new MyPromise((resolve, reject) => {
    //   resolve('success')
    // })
    // p2 = p1.then('这里的onFulfilled是一个文本' )
    // p2.then(
    //   res => { console.log(res) },
    //   err => {console.log(err) }
    // )  //success

    // 例子4:
    // p1 = new MyPromise((resolve, reject) => {
    //   reject('fail')
    // })
    // p2 = p1.then(res => {
    //  return 'sucess'
    // }, '这里的onRejected是一个文本')
    // p2.then(
    //   res => { console.log(res)},
    //   err => { console.log(err)}
    // )  //fail

  </script>
</body>

</html>
